cmd/restic/cmd_ls.go

```diff
@@ -59,20 +74,89 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
 		return err
 	}
 
+	// extract any specific directories to walk
+	var dirs []string
+	if len(args) > 1 {
+		dirs = args[1:]
+		for _, dir := range dirs {
+			if !strings.HasPrefix(dir, "/") {
+				return errors.Fatal("All path filters must be absolute, starting with a forward slash '/'")
+			}
+		}
+	}
+
 	ctx, cancel := context.WithCancel(gopts.ctx)
 	defer cancel()
-	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args) {
-		Verbosef("snapshot %s of %v at %s):\n", sn.ID().Str(), sn.Paths, sn.Time)
+	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args[:1]) {
+		Verbosef("snapshot %s of %v filtered by %v at %s):\n", sn.ID().Str(), sn.Paths, dirs, sn.Time)
 
 		err := walker.Walk(ctx, repo, *sn.Tree, nil, func(nodepath string, node *restic.Node, err error) (bool, error) {
 			if err != nil {
 				return false, err
 			}
-
 			if node == nil {
 				return false, nil
 			}
+
+			// apply any directory filters
+			if len(dirs) > 0 {
+				nodeDir := path.Dir(nodepath)
+
+				// this first iteration ensures we do not traverse branches that
+				// are not in matching trees or will not lead us to matching trees
+				var walk bool
+				for _, dir := range dirs {
+					approachingMatchingTree := fs.HasPathPrefix(nodeDir, dir)
+					inMatchingTree := fs.HasPathPrefix(dir, nodepath)
+
+					// this condition is complex, but it basically requires that we
+					// are either approaching a matching tree (not yet deep enough)
+					// or: if recursive, we have entered a matching tree; if non-
+					// recursive, then that we are at exactly the right depth
+					// (we can do the walk correctly by just using the condition of
+					// "approachingMatchingTree || inMatchingTree", but it will be
+					// much slower for non-recursive queries since it will continue
+					// to traverse subtrees that are too deep and won't match -- this
+					// extra check allows us to return SkipNode if we've gone TOO deep,
+					// which skips all its subfolders)
+					if approachingMatchingTree || opts.Recursive || (inMatchingTree && dir == nodeDir) {
+						walk = true
+						break
+					}
+				}
+				if !walk {
+					return false, walker.SkipNode
```

Comment:
We need to be careful here: When we return `SkipNode` for a node that is not a dir (e.g. a file or other type), then the remaining nodes in the same tree are skipped!